### PR TITLE
Add `Ast_pattern.esequence`

### PR DESCRIPTION
**Pull request by matthewelse**
This PR adds support for matching on a list of subsequent expressions that form a sequence, as a natural analogue to `Ast_builder.esequence`.

### CODE REVIEW

File: src/ast_pattern.ml [hunk 1]
```diff
@@ -220,5 +220,15 @@ let elist (T f) = T (fun ctx _loc e k ->
   k (List.map l ~f:(fun x -> f ctx x.Parsetree.pexp_loc x (fun x -> x))))
 ;;
 
+let esequence (T f) = T (fun ctx _loc e k ->
+  let rec parse_seq expr =
+    match expr.pexp_desc with
+    | Pexp_sequence (expr, next) ->
+      expr :: (parse_seq next)
+    | _ -> [ expr ]
```
Comment (pitag-ha) [inline src/ast_pattern.ml:228]: Do you think it might be worth making a tail recursion out if this?

**Comment by pitag-ha** [comment]
Do you have a concrete use case for this in mind?

**Comment by matthewelse** [comment]
A colleague of mine wanted this feature - I don't know exactly what they were trying to do, but it seemed like a simple enough thing to include in ppxlib itself

**Comment by pitag-ha** [comment]
Do you know why your colleague can't use `elist` / prefers `esequence` over `elist`? Do they want to make sure that the first `n-1` expressions return `unit`?

**Comment by matthewelse** [comment]
`elist` matches a list, right?

```
let x = [1; 2; 3]
```

I'm trying to match the following:

```
let () =
  do_first ();
  do_second ();
  do_third ()
```

As far as I can tell, there's no way of getting a list of expressions out of the second example without `esequence`, or some hand written code.

**Comment by pitag-ha** [comment]
Ok, I can see now why to choose that syntax instead of a list. Thanks! To me adding this feature sounds good.

Have you seen [this comment](https://github.com/ocaml-ppx/ppxlib/pull/264#discussion_r658905374)?

File: src/ast_pattern.ml (hunk 1 above)
Comment (pitag-ha) [reply src/ast_pattern.ml:228]: Similarly to [here](https://github.com/ocaml-ppx/ppxlib/blob/master/src/ast_pattern.ml#L209).

**Comment by matthewelse** [comment]
Thanks @pitag-ha. I think this should now resolve the comment.

**Review by pitag-ha — APPROVED** [review]
Thanks @matthewelse! LGTM. Since I'm not very familiar with `Ast_pattern`, I'll wait some days to merge this in case @jeremiedimino or @NathanReb want to say something about it before.

**Comment by matthewelse** [comment]
Thanks, sounds great!

**Comment by jeremiedimino** [comment]
`esequence` seems like a good addition to me

**Pull request merged**